### PR TITLE
Add WaitForAck and LivelinessTest Tests To TSAN Coverage

### PR DESCRIPTION
**Pull request by simpsont-oci**
Add a few more integration tests with different code paths to TSAN coverage. Fix things until TSAN complaining.

### CODE REVIEW

File: dds/DCPS/DataReaderImpl.cpp [hunk 1]
```diff
@@ -2141,9 +2134,11 @@ std::ostream& OpenDDS::DCPS::WriterStats::raw_data(std::ostream& str) const
 void
 DataReaderImpl::writer_removed(WriterInfo& info)
 {
+  PublicationId info_writer_id = info.writer_id();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2137]:
```suggestion
  const PublicationId& info_writer_id = info.writer_id();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 2]
```diff
@@ -2164,20 +2159,22 @@ DataReaderImpl::writer_removed(WriterInfo& info)
   {
     ACE_GUARD(ACE_Recursive_Thread_Mutex, guard, sample_lock_);
 
-    if (info.state_ == WriterInfo::ALIVE) {
+    WriterInfo::WriterState info_state = info.state();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2162]:
```suggestion
    const WriterInfo::WriterState& info_state = info.state();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 3]
```diff
@@ -2190,9 +2187,11 @@ void
 DataReaderImpl::writer_became_alive(WriterInfo& info,
     const MonotonicTimePoint& /* when */)
 {
+  PublicationId info_writer_id = info.writer_id();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2190]:
```suggestion
  const PublicationId& info_writer_id = info.writer_id();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 4]
```diff
@@ -2207,19 +2206,21 @@ DataReaderImpl::writer_became_alive(WriterInfo& info,
 
   bool liveliness_changed = false;
 
-  if (info.state_ != WriterInfo::ALIVE) {
+  WriterInfo::WriterState info_state = info.state();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2209]:
```suggestion
  const WriterInfo::WriterState& info_state = info.state();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 5]
```diff
@@ -2263,9 +2264,11 @@ DataReaderImpl::writer_became_alive(WriterInfo& info,
 void
 DataReaderImpl::writer_became_dead(WriterInfo& info)
 {
+  PublicationId info_writer_id = info.writer_id();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2267]:
```suggestion
  const PublicationId& info_writer_id = info.writer_id();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 6]
```diff
@@ -2277,32 +2280,34 @@ DataReaderImpl::writer_became_dead(WriterInfo& info)
...
+    owner_manager->remove_writer(info_writer_id);
+    info.clear_owner_evaluated();
   }
 #endif
 
   // caller should already have the sample_lock_ !!!
   bool liveliness_changed = false;
 
-  if (info.state_ == OpenDDS::DCPS::WriterInfo::NOT_SET) {
+  WriterInfo::WriterState info_state = info.state();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:2291]:
```suggestion
  const WriterInfo::WriterState& info_state = info.state();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 7]
```diff
@@ -3005,40 +3010,43 @@ DataReaderImpl::update_ownership_strength (const PublicationId& pub_id,
 #endif
 
 #ifndef OPENDDS_NO_OBJECT_MODEL_PROFILE
-bool DataReaderImpl::verify_coherent_changes_completion (WriterInfo* writer)
+bool DataReaderImpl::verify_coherent_changes_completion(WriterInfo* writer)
 {
   Coherent_State state = COMPLETED;
   bool accept_here = true;
 
+  PublicationId writer_id = writer->writer_id();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:3018]:
```suggestion
  const PublicationId& writer_id = writer->writer_id();
```

File: dds/DCPS/DataReaderImpl.cpp [hunk 8]
```diff
@@ -3005,40 +3010,43 @@ DataReaderImpl::update_ownership_strength (const PublicationId& pub_id,
...
+  RepoId publisher_id = writer->publisher_id();
```
Comment (jrw972) [inline dds/DCPS/DataReaderImpl.cpp:3019]:
```suggestion
  const RepoId& publisher_id = writer->publisher_id();
```

File: dds/DCPS/DataWriterImpl.cpp [hunk 9]
```diff
@@ -387,6 +378,17 @@ DataWriterImpl::association_complete_i(const RepoId& remote_id)
   {
     ACE_GUARD(ACE_Recursive_Thread_Mutex, guard, this->lock_);
 
+    if (DCPS_debug_level >= 1) {
+      GuidConverter writer_converter(this->publication_id_);
```
Comment (jrw972) [inline dds/DCPS/DataWriterImpl.cpp:382]: Use LogGuid.

**Review by jrw972 — CHANGES_REQUESTED** [review]
I think that most of my "capture references to repo ids" should be ignored.

File: dds/DCPS/WriterInfo.h [hunk 10]
```diff
@@ -98,33 +101,132 @@ class OpenDDS_Dcps_Export WriterInfo : public RcObject {
   void received_activity(const MonotonicTimePoint& when);
 
   /// returns 1 if the DataWriter is lively; 2 if dead; otherwise returns 0.
-  WriterState get_state() {
+  WriterState state() const {
+    ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
     return state_;
   };
```
Comment (mitza-oci) [inline dds/DCPS/WriterInfo.h:107]:
```suggestion
  WriterState state() const
  {
    ACE_Guard<ACE_Thread_Mutex> guard(mutex_);
    return state_;
  }
```
more like this

File: dds/DCPS/transport/framework/TransportClient.h [hunk 11]
```diff
@@ -127,7 +127,7 @@ class OpenDDS_Dcps_Export TransportClient
   bool remove_all_msgs();
 
   virtual void add_link(const DataLink_rch& link, const RepoId& peer);
-  virtual const RepoId& get_repo_id() const = 0;
+  virtual const RepoId get_repo_id() const = 0;
```
Comment (mitza-oci) [inline dds/DCPS/transport/framework/TransportClient.h:130]:
```suggestion
  virtual RepoId get_repo_id() const = 0;
```

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**